migrations/admin_events.go

```diff
@@ -0,0 +1,151 @@
+	User         string           `bson:"user"`
+	Section      string           `bson:"section"`
+	Changes      configDataChange `bson:"changes"`
+}
+
+type configDataChange struct {
+	Before interface{} `bson:"before"`
+	After  interface{} `bson:"after"`
+}
+
+type dbSettings struct {
+	Banner      string `bson:"banner"`
+	BannerTheme string `bson:"banner_theme"`
+}
+
+func adminEventRestructureGenerator(env anser.Environment, dbName string, limit int) (anser.Generator, error) {
+	const migrationName = "admin_event_restructure"
+
+	if err := env.RegisterManualMigrationOperation(migrationName, makeAdminEventMigration(dbName)); err != nil {
+		return nil, err
+	}
+
+	opts := model.GeneratorOptions{
+		NS: model.Namespace{
+			DB:         dbName,
+			Collection: eventCollection,
+		},
+		Limit: limit,
+		Query: db.Document{
+			"r_id":        "",
+			"data.r_type": adminDataType,
+		},
+		JobID: "migration-admin-event-restructure",
+	}
+
+	return anser.NewManualMigrationGenerator(env, opts, migrationName), nil
+}
+
+func makeAdminEventMigration(database string) db.MigrationOperation {
+	const (
+		idKey   = "_id"
+		dataKey = "data"
+	)
+
+	return func(session db.Session, rawD bson.RawD) error {
+		defer session.Close()
+
+		var docId bson.ObjectId
+		oldData := eventDataOld{}
+		changeType := ""
+		for _, raw := range rawD {
+			switch raw.Name {
+			case idKey:
+				if err := raw.Value.Unmarshal(&docId); err != nil {
+					return errors.Wrap(err, "error unmarshaling id")
+				}
+			case eventTypeKey:
+				if err := raw.Value.Unmarshal(&changeType); err != nil {
+					return errors.Wrap(err, "error unmarshaling event type")
+				}
+			case dataKey:
+				if err := raw.Value.Unmarshal(&oldData); err != nil {
+					return errors.Wrap(err, "error unmarshaling event data")
+				}
+			}
+		}
+		if changeType == "" {
+			return errors.New("change type is empty")
+		}
+
+		newData := eventDataNew{
+			ResourceType: oldData.ResourceType,
+			User:         oldData.User,
+		}
+		switch changeType {
+		case eventTypeTheme:
+			before := &dbSettings{BannerTheme: oldData.OldVal}
+			after := &dbSettings{BannerTheme: oldData.NewVal}
+			newData.Section = "global"
+			newData.Changes = configDataChange{Before: before, After: after}
+		case eventTypeBanner:
+			before := &dbSettings{Banner: oldData.OldVal}
+			after := &dbSettings{Banner: oldData.NewVal}
+			newData.Section = "global"
+			newData.Changes = configDataChange{Before: before, After: after}
+		case eventTypeServiceFlags:
+			before := oldData.OldFlags
+			after := oldData.NewFlags
+			newData.Section = "service_flags"
+			newData.Changes = configDataChange{Before: &before, After: &after}
+		default:
+			return fmt.Errorf("unexpected change type %s found", changeType)
+		}
+
+		return session.DB(database).C(eventCollection).UpdateId(docId,
+			db.Document{
+				"$set": db.Document{
+					dataKey:      newData,
+					eventTypeKey: eventTypeValueChanged,
+				},
```

Comment:
You're also going to want to unset old_val, new_val, old_flags, new_flags.

Reply:
This isn't needed since I'm replacing the entire data sub-document, but I added test cases for this